File: src/Avalonia.Samples/CustomControls/RatingControlSample/ViewLocator.cs
using Avalonia.Controls;
using Avalonia.Controls.Templates;
using RatingControlSample.ViewModels;
using System;

namespace RatingControlSample
{
    public class ViewLocator : IDataTemplate
    {
        public Control Build(object? data)
        {
            var name = data?.GetType().FullName!.Replace("ViewModel", "View") ?? "<null>";
            var type = Type.GetType(name);

            if (type != null)
            {
                return (Control)Activator.CreateInstance(type)!;
            }
            else
            {
                return new TextBlock { Text = "Not Found: " + name };
            }
        }

        public bool Match(object data)
        {
            return data is ViewModelBase;
        }
    }
}

File: src/Avalonia.Samples/FSharp/CompleteApps/MusicStore/ViewLocator.fs
namespace MusicStore

open System
open Avalonia.Controls
open Avalonia.Controls.Templates
open MusicStore.ViewModels

type ViewLocator() =
    interface IDataTemplate with

        member this.Build(data) =
            if isNull data then
                null
            else
                let name = data.GetType().FullName.Replace("ViewModel", "View", StringComparison.Ordinal)
                let typ = Type.GetType(name)
                if isNull typ then
                    upcast TextBlock(Text = $"Not Found: %s{name}")
                else
                    downcast Activator.CreateInstance(typ)

        member this.Match(data) = data :? ViewModelBase